[ClangImporter] Don't import C structs with non-trivial fields

jrose-apple opened the pull request:
Swift currently assumes that copies and deinitializations of C structs are always "trivial", i.e. a copy can be done by just copying the raw bits of the value, and a deinit requires no work at all. Until we do anything to change that, just don't import any of the new C structs that have non-trivial copy or deinitialization operations (those with ARC references in them).

rdar://problem/35698831

jrose-apple commented:
Blocked on \<rdar://problem/38421774\> NonTrivialToCopy et al not preserved in serialized AST (PCH/PCM).

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3068,6 +3068,25 @@ namespace {
+      // FIXME: We should actually support strong ARC references and similar in
+      // C structs. Unfortunately that might mean implementing our own copy of
+      // these predicates, because Swift might only support a subset of what
+      // Clang does.
```
We would not need to implement our own copy of these predicates.  We might be unable to import all the fields, but that shouldn't prevent us from importing the struct types themselves; we could even support adding initializers by implicitly default-initializing any fields that aren't importable.  None of this is conceptually difficult.  The only reason we need to disable this for now is that SIL type lowering doesn't know to mark C structs address-only and IRGen doesn't know how to ask Clang to copy/destroy/default-initialize a non-trivial struct.

And in many cases we would not be limited from importing the fields.  ObjC's ``__weak`` is not our native ``weak``, but we could certainly add a reference storage type for it (mostly for SIL's purposes) and otherwise ignore the qualifier in the AST type system.

jrose-apple replied:
That's one possible implementation strategy. Another would be to use Swift's logic to copy the structs field-by-field, although that only works if all of the field types are representable in Swift.

rjmccall replied:
Right.  The point is that the decision of how to copy the type is clearly just an implementation detail of IRGen and shouldn't ultimately limit what we can import.

IRGen is going to have to be able to implement copies and destroys by calling into Clang-generated code in order to eventually handle C++ types, and the optimizer issues with that are very similar to the issues with unimported fields (i.e. it's not okay to destructure the struct), so there are technical motivations pushing us to handle things that way.

jrose-apple replied:
Okay. Do you think I should change the comment, or is it fine to leave it?

rjmccall replied:
It looks like you have to resubmit anyway, so would you mind changing it?  I think it would be okay to explain *less* and just say that we should support these things but that doing so will require some work in SIL and IRGen.

jrose-apple commented:
Failure on the "source compat" bot should hit the regular macOS bot too: it looks like the Clang-side patch changes the meaning of unions, which it shouldn't.

jrose-apple commented:
(but there's also a problem with anonymous fields here that's independent)

jrose-apple commented:
Okay, maybe the anonymous field assertion is valid. The only reasons a record declaration should fail to import are ones that can't apply to anonymous fields.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
(still planning to update the comment before merging)

jrose-apple committed:
[ClangImporter] Don't import C structs with non-trivial fields

Swift currently assumes that copies and deinitializations of C structs
are always "trivial", i.e. a copy can be done by just copying the raw
bits of the value, and a deinit requires no work at all. Until we do
anything to change that, just don't import any of the new C structs
that have non-trivial copy or deinitialization operations (those with
ARC references in them).

rdar://problem/35698831

jrose-apple commented:
@swift-ci Please smoke test
